gnovm/pkg/gnolang/internal/txlog/txlog.go

```diff
@@ -0,0 +1,141 @@
+// Package txlog is an internal package containing data structures that can
+// function as "transaction logs" on top of a hash map (or other key/value
+// data type implementing [Map]).
+//
+// A transaction log keeps track of the write operations performed in a
+// transaction, so that they can be committed together, atomically,
+// when calling [MapCommitter.Commit].
+package txlog
+
+// Map is a generic interface to a key/value map, like Go's builtin map.
+type Map[K comparable, V any] interface {
+	Get(K) (V, bool)
+	Set(K, V)
+	Delete(K)
+	Iterate() func(yield func(K, V) bool)
+}
+
+// MapCommitter is a Map which also implements a Commit() method, which writes
+// to the underlying (parent) [Map].
+type MapCommitter[K comparable, V any] interface {
+	Map[K, V]
+
+	// Commit writes the logged operations to the underlying map.
+	// After calling commit, the underlying tx log is cleared and the
+	// MapCommitter may be reused.
+	Commit()
+}
+
+// GoMap is a simple implementation of [Map], which wraps the operations of
+// Go's map builtin to implement [Map].
+type GoMap[K comparable, V any] map[K]V
+
+// Get implements [Map].
+func (m GoMap[K, V]) Get(k K) (V, bool) {
+	v, ok := m[k]
+	return v, ok
+}
+
+// Set implements [Map].
+func (m GoMap[K, V]) Set(k K, v V) {
+	m[k] = v
+}
+
+// Delete implements [Map].
+func (m GoMap[K, V]) Delete(k K) {
+	delete(m, k)
+}
+
+// Iterate implements [Map].
+func (m GoMap[K, V]) Iterate() func(yield func(K, V) bool) {
+	return func(yield func(K, V) bool) {
+		for k, v := range m {
+			if !yield(k, v) {
+				return
+			}
+		}
+	}
+}
+
+// Wrap wraps the map m into a data structure to keep a transaction log.
+// To write data to m, use MapCommitter.Commit.
+func Wrap[K comparable, V any](m Map[K, V]) MapCommitter[K, V] {
+	return &txLog[K, V]{
+		source: m,
+		dirty:  make(map[K]deletable[V]),
+	}
+}
+
+type txLog[K comparable, V any] struct {
+	source Map[K, V]          // read-only until Commit()
+	dirty  map[K]deletable[V] // pending writes on source
+}
+
+func (b *txLog[K, V]) Commit() {
```

Comment:
I'm not sure if this should be named 'Commit.'   It does not commit anything.

Reply:
You're welcome to make a PR to propose a rename to `Write` - or anything else

Reply:
(My reasoning here for the name comes from the terminology also used in SQL; which uses BEGIN, ROLLBACK and COMMIT)